### PR TITLE
[Fix] fix local_visualizer.py: only visualize when there is at least one instance

### LINKED ISSUE #2119: [Bug] _draw_instances_3d crashes when no detection
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection3d/issues) and [Discussions](https://github.com/open-mmlab/mmdetection3d/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://github.com/open-mmlab/mmdetection3d/blob/1.1/docs/en/notes/faq.md) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (dev)](https://github.com/open-mmlab/mmdetection3d/tree/dev) or [latest version (1.x)](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x).

### Task

I'm using the official example scripts/configs for the officially supported tasks/models/datasets.

### Branch

1.1x branch https://github.com/open-mmlab/mmdetection3d/tree/1.1

### Environment

sys.platform: linux
Python: 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:49:35) [GCC 10.4.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA T1200 Laptop GPU
CUDA_HOME: /home/mambaforge/envs/openmmlab
NVCC: Cuda compilation tools, release 11.7, V11.7.99
GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 1.13.0
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2022.1-Product Build 20220311 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.7
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.5
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.7, CUDNN_VERSION=8.5.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.14.0
OpenCV: 4.6.0
MMEngine: 0.3.2
MMDetection: 3.0.0rc4
MMDetection3D: 1.1.0rc2+7fd57b1
spconv2.0: False

### Reproduces the problem - code sample

NA

### Reproduces the problem - command or script

```shell
python ./tools/test.py ./configs/pointpillars/debug_train_config.py ./work_dirs/debug_train_config/epoch_80.pth --show
```

### Reproduces the problem - error message

```
  File "/home/workspace/mmdetection3d-1.1/mmdet3d/visualization/local_visualizer.py", line 731, in add_datasample
    gt_data_3d = self._draw_instances_3d(
  File "/home/workspace/mmdetection3d-1.1/mmdet3d/visualization/local_visualizer.py", line 564, in _draw_instances_3d
    bboxes_3d = instances.bboxes_3d  # BaseInstance3DBoxes
AttributeError: 'InstanceData' object has no attribute 'bboxes_3d'
```

### Additional information

Code below fails when there is no detection 

https://github.com/open-mmlab/mmdetection3d/blob/d7067e4430f4e99697b382b0e3ea5597af737a2c/mmdet3d/visualization/local_visualizer.py#L543-L560

### PR DESCRIPTION
**Pull request by 404Vector**
- fix [AttributeError: ‘InstanceData’ object has no attribute ‘bboxes_3d’] error

**An error occurs when drawing a 3d box on an image to evaluate model performance as shown below.**

`python tools/test.py configs/smoke/smoke_dla34_dlaneck_gn-all_4xb8-6x_kitti-mono3d.py checkpoints/smoke/smoke_dla34_pytorch_dlaneck_gn-all_8x4_6x_kitti-mono3d_20210929_015553-d46d9bb0.pth --show --show-dir='work_dirs/test' --task='mono_det'`

## Motivation

I wanted to draw a 3D BBox using tools/test.py.
However, it could not be drawn due to an internal error.
I'll fix the error with this PR.

## Modification
file : mmdet3d/visualization/local_visualizer.py
 > class : Det3DLocalVisualizer Class
 >> method : add_datasample, _draw_instances_3d

Looking at the data_sample object at runtime, it has gt_instances_3d.
But inside gt_instances_3d the data doesn't exist.
Fix the bug by modifying the code as shown below.

* before
```
if 'gt_instances_3d' in data_sample:
    gt_data_3d = self._draw_instances_3d(
          data_input, data_sample.gt_instances_3d,
          data_sample.metainfo, vis_task, palette)
```

* after
```
if 'gt_instances_3d' in data_sample:
     if 'bboxes_3d' in data_sample.gt_instances_3d:
          gt_data_3d = self._draw_instances_3d(
               data_input, data_sample.gt_instances_3d,
               data_sample.metainfo, vis_task, palette)
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

fix #2119

### CODE REVIEW

File: mmdet3d/visualization/local_visualizer.py [hunk 1]
```diff
@@ -723,9 +723,10 @@ def add_datasample(self,
 
         if draw_gt and data_sample is not None:
             if 'gt_instances_3d' in data_sample:
-                gt_data_3d = self._draw_instances_3d(
-                    data_input, data_sample.gt_instances_3d,
-                    data_sample.metainfo, vis_task, palette)
+                if 'bboxes_3d' in data_sample.gt_instances_3d:
```
Comment (Xiangxu-0103) [inline mmdet3d/visualization/local_visualizer.py:726]:
```suggestion
                if len(data_sample.gt_instances_3d) > 0:
```

Comment (Xiangxu-0103) [reply mmdet3d/visualization/local_visualizer.py:726]: Please also add for `pred_instances_3d`

**Comment by Xiangxu-0103** [comment]
Please pull the latest code to fix lint issue and change your base to `dev-1.x` branch

File: mmdet3d/visualization/local_visualizer.py (hunk 1 above)
Comment (404Vector) [reply mmdet3d/visualization/local_visualizer.py:726]: I got it! :)

**Comment by 404Vector** [comment]
@Xiangxu-0103

- [ ] Check for conflicts between the latest version (dev-1.x) and my work
- [ ] Merge the dev-1.x branch into my branch
- [ ] Apply suggested change and Check if the code is working properly
- [ ] Add Code for pred_instances_3d and Check if the code is working properly
- [ ] Pre-commit

**Comment by Xiangxu-0103** [comment]
download
Please change your merge branch to `dev-1.x`

**Review by Xiangxu-0103 — APPROVED** [review]

**Review by Tai-Wang — APPROVED** [review]

**Review by ZCMax — APPROVED** [review]

File: mmdet3d/visualization/local_visualizer.py [hunk 2]
```diff
@@ -747,15 +748,15 @@ def add_datasample(self,
...
-                                                       pred_instances_3d,
-                                                       data_sample.metainfo,
-                                                       vis_task, palette)
+                if len(data_sample.pred_instances_3d) > 0:
+                    pred_instances_3d = data_sample.pred_instances_3d
+                    # .cpu can not be used for BaseInstancesBoxes3D
+                    # so we need to use .to('cpu')
+                    pred_instances_3d = pred_instances_3d[
+                        pred_instances_3d.scores_3d > pred_score_thr].to('cpu')
+                    pred_data_3d = self._draw_instances_3d(
```
Comment (ZwwWayne) [inline mmdet3d/visualization/local_visualizer.py:757]: It is better to do the check and skip drawing in self._draw_instances_3d to reduce and avoid the modification in two places.

Comment (404Vector) [reply mmdet3d/visualization/local_visualizer.py]: The exception handling location has been changed to reflect your opinion.

Thank you for your kind comments.

**Comment by 404Vector** [comment]
@Xiangxu-0103 Please approve the circle ci work flow :)

**Comment by jason102811** [comment]
Hi @404Vector  ！First of all, we want to express our gratitude for your significant PR in the MMDet. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**